### PR TITLE
[Issue #3436][pulsar-broker] Creating REST Endpoint for non-partitioned topic creation

### PR DESCRIPTION
**Pull request by ConcurrencyPractitioner**
We are adding a REST endpoint which allows the admin to create non-partitioned topics
through PersistentTopics.

### CODE REVIEW

**Comment by ConcurrencyPractitioner** [comment]
Hi @sijie I am slightly confused on how to create this REST endpoint. A non-partitioned topic is created in ```internalCreateSubscription```, but functions as a temporary topic from which a subscription could be created. So if we created a PersistentTopic outside of this method, what would we do with it?

**Comment by ConcurrencyPractitioner** [comment]
In our previous discussion on the issue, I remember that ```createPartitionedTopic``` needed to be improved. I noticed that there is a precondition which prevents the number of partitions from being created to be less than two. So my understanding here is that we should remove this condition (i.e. allow the creation of a topic with only one partition) ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -398,6 +398,18 @@ protected void internalCreatePartitionedTopic(int numPartitions, boolean authori
         }
     }
 
+    protected void internalCreateNonPartitionedTopic(boolean authoritative) {
+    	validateAdminAccessForTenant(topicName.getTenant());
+    	
+    	try {
+    		PersistentTopic topic = (PersistentTopic) getOrCreateTopic(topicName);
+    		//...
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: what is this for?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 2]
```diff
@@ -140,6 +140,23 @@ public void createPartitionedTopic(@PathParam("tenant") String tenant, @PathPara
...
+    @ApiResponses(value = {
+        @ApiResponse(code = 403, message = "Don't have admin permission"),
+        @ApiResponse(code = 409, message = "Partitioned topic already exist"),
+        @ApiResponse(code = 412, message = "Failed Reason : Name is invalid or Namespace does not have any clusters configured"),
+        @ApiResponse(code = 503, message = "Failed to validate global cluster configuration")
+    })
+    public void createNonPartitionedTopic(@PathParam("tenant") String tenant, @PathParam("namespace") String namespace,
+            @PathParam("topic") @Encoded String encodedTopic, 
+            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative) {
+    	validateGlobalNamespaceOwnership(tenant,namespace);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java]: nit: wrong indent

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 3]
```diff
@@ -140,6 +140,23 @@ public void createPartitionedTopic(@PathParam("tenant") String tenant, @PathPara
...
+        validatePartitionedTopicName(tenant, namespace, encodedTopic);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java]: this should not be `validatePartitionedTopicName`

**Comment by sijie** [comment]
> A non-partitioned topic is created in `internalCreateSubscription`, but functions as a temporary topic from which a subscription could be created. So if we created a non-partitioned PersistentTopic outside of this method, what would we do with it?

Currently the default behavior is a non-partitioned topic automatically created when a consumer subscribe or a publisher publishes a message. With you previous change, we disable this auto-creation. so the endpoint is the only way to create the non-partitioned topic. If a topic is created by the rest endpoint, when consumer/publisher connect, they will be realizing that the topic has been created. they will continue consuming or publishing messages.

> I remember that `createPartitionedTopic` needed to be improved. I noticed that there is a precondition which prevents the number of partitions from being created to be less than two. So my understanding here is that we should remove this condition (i.e. allow the creation of a topic with no partitions) ?

I think that should be a different pull request and a separated discussion. Changing the behavior of creating partitioned topic (e.g. allow creating a partitioned topic with only 1 partition) should be discussed at dev@ mailing list. It should be done via a PIP, since it is a user facing behavior change though.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (ConcurrencyPractitioner) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: Sorry about that, that was meant to denote that there should be some lines of code below it. (which I will probably add later)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 3 above)
Comment (ConcurrencyPractitioner) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java]: Yeah, good point.

**Comment by ConcurrencyPractitioner** [comment]
@sijie Ah, I see. I didn't realize that. I will add tests ASAP.

**Comment by ConcurrencyPractitioner** [comment]
@sijie By the way, for this test that we are to add, how would we test it? Should it be somewhat similar like the last test I added (i.e. it would be in ```PersistentTopicsTest```) or should it be somewhere else (like ```AdminApiTest```)?

**Comment by ConcurrencyPractitioner** [comment]
@sijie Do you think there should be anything else?

**Comment by ConcurrencyPractitioner** [comment]
Actually. My thinking was that since these PRs are pretty small, we could just commit the changes on top of this one. We could avoid the hassle of moving our conversations in between PRs.

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

**Comment by ConcurrencyPractitioner** [comment]
rerun java 8 - unit tests

**Comment by ConcurrencyPractitioner** [comment]
run java 8 - unit tests

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

**Comment by sijie** [comment]
run java8 tests

**Comment by ConcurrencyPractitioner** [comment]
@sijie I guess this could be merged  unless you think there should be something else that should be done.

**Comment by sijie** [comment]
run java8 tests

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java [hunk 4]
```diff
@@ -82,6 +82,7 @@ public CmdTopics(PulsarAdmin admin) {
         jcommander.addCommand("expire-messages-all-subscriptions", new ExpireMessagesForAllSubscriptions());
 
         jcommander.addCommand("create-partitioned-topic", new CreatePartitionedCmd());
+        jcommander.addCommand("create-non-partitioned-topic", new CreateNonPartitionedCmd());
```
Comment (cckellogg) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java]: what about just this for the cli?
`pulsar-admin topics create my-topic`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 5]
```diff
@@ -140,6 +140,22 @@ public void createPartitionedTopic(@PathParam("tenant") String tenant, @PathPara
         internalCreatePartitionedTopic(numPartitions, authoritative);
     }
 
+    @PUT
+    @Path("/{tenant}/{namespace}/{topic}/nonpartitioned")
```
Comment (cckellogg) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java]: What about just /{tenant}/{namespace}/{topic} for the endpoint to create? It seems more natural and the /nonpartitioned is kind of confusing and seems inconsistent with the rest of the endpoints for topics. For example we do /{tenant}/{namespace}/{topic}/stats to get the stats and not /{tenant}/{namespace}/{topic}/nonpartitioned-stats

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java (hunk 4 above)
Comment (ConcurrencyPractitioner) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java]: Sure, I was just following the standard for create-partitioned-topic. But I guess we could change that.

**Comment by jiazhai** [comment]
run integration tests

**Comment by ConcurrencyPractitioner** [comment]
Retest this please.

**Comment by ConcurrencyPractitioner** [comment]
Jenkins: Integration tests seem to have some problems (related to Docker Images, I don't know how).

**Comment by jiazhai** [comment]
run java8 tests
run integration tests

**Comment by ConcurrencyPractitioner** [comment]
@jiazhai or @sijie If you feel there is anything else that should be added, let me know. 👍

**Comment by ConcurrencyPractitioner** [comment]
@sijie @jiazhai or @cckellogg I guess this could be merged then.

**Comment by sijie** [comment]
@ConcurrencyPractitioner will merge when the Ci passed.

**Comment by sijie** [comment]
run java8 tests

**Comment by ConcurrencyPractitioner** [comment]
run java8 tests

**Comment by ConcurrencyPractitioner** [comment]
run java8 tests

**Comment by ConcurrencyPractitioner** [comment]
Ok, so I ran the failing test from Jenkins (```BrokerServiceLookupTest#testWebserviceServiceTls```) on local with my change. It passed, so I guess it's just flaking.

**Comment by ConcurrencyPractitioner** [comment]
run java8 tests

**Pull request merged**